cli/command/stack/kubernetes/stackclient.go

```diff
@@ -0,0 +1,155 @@
+	"k8s.io/client-go/rest"
+)
+
+// stackClient talks to a kubernetes compose component.
+type stackClient interface {
+	CreateOrUpdate(s stack) error
+	Delete(name string) error
+	Get(name string) (stack, error)
+	List(opts metav1.ListOptions) ([]stack, error)
+	IsColliding(servicesClient corev1.ServiceInterface, s stack) error
+}
+
+// stackV1Beta1 implements stackClient interface and talks to compose component v1beta1.
+type stackV1Beta1 struct {
+	stacks composev1beta1.StackInterface
+}
+
+func newStackV1Beta1(config *rest.Config, namespace string) (stackClient, error) {
+	client, err := composev1beta1.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &stackV1Beta1{stacks: client.Stacks(namespace)}, nil
+}
+
+func (s *stackV1Beta1) CreateOrUpdate(internalStack stack) error {
+	// If it already exists, update the stack
+	if stackBeta1, err := s.stacks.Get(internalStack.name, metav1.GetOptions{}); err == nil {
+		stackBeta1.Spec.ComposeFile = internalStack.composeFile
+		_, err := s.stacks.Update(stackBeta1)
+		return err
+	}
+	// Or create it
+	_, err := s.stacks.Create(stackToV1beta1(internalStack))
+	return err
+}
+
+func (s *stackV1Beta1) Delete(name string) error {
+	return s.stacks.Delete(name, &metav1.DeleteOptions{})
+}
+
+func (s *stackV1Beta1) Get(name string) (stack, error) {
+	stackBeta1, err := s.stacks.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return stack{}, err
+	}
+	return stackFromV1beta1(stackBeta1)
+}
+
+func (s *stackV1Beta1) List(opts metav1.ListOptions) ([]stack, error) {
+	list, err := s.stacks.List(opts)
+	if err != nil {
+		return nil, err
+	}
+	stacks := make([]stack, len(list.Items))
+	for i := range list.Items {
+		stack, err := stackFromV1beta1(&list.Items[i])
+		if err != nil {
+			return nil, err
+		}
+		stacks[i] = stack
+	}
+	return stacks, nil
+}
+
+// IsColliding verifies that services defined in the stack collides with already deployed services
+func (s *stackV1Beta1) IsColliding(servicesClient corev1.ServiceInterface, st stack) error {
+	for _, srv := range st.getServices() {
+		if err := verify(servicesClient, st.name, srv); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// verify checks wether the service is already present in kubernetes.
+// If we find the service by name but it doesn't have our label or it has a different value
+// than the stack name for the label, we fail (i.e. it will collide)
+func verify(services corev1.ServiceInterface, stackName string, service string) error {
+	svc, err := services.Get(service, metav1.GetOptions{})
+	if err == nil {
+		if key, ok := svc.ObjectMeta.Labels[labels.ForStackName]; ok {
+			if key != stackName {
+				return fmt.Errorf("service %s already present in stack named %s", service, key)
+			}
+			return nil
+		}
+		return fmt.Errorf("service %s already present in the cluster", service)
+	}
+	return nil
+}
+
+// stackV1Beta2 implements stackClient interface and talks to compose component v1beta2.
```

Comment:
nit: tempted to say; move the `stackV1Beta1` and `stackV1Beta2` bits to separate files